### PR TITLE
feat(coderd): add ability to mark workspaces as favorite

### PR DESCRIPTION
**Pull request by johnstcn**
Part of https://github.com/coder/coder/issues/7912

- [x] Update DB schema + queries (this PR)
- [x] Add API endpoints to favorite/unfavorite workspace (this PR)
- [x] Show favorite status in UI/CLI (follow-up)
- [x] Add CLI commands to favorite/unfavorite workspace (follow-up)
- [x] Add UI support to favorite/unfavorite workspace (follow-up)

**Goal:**

When you query workspaces in the UI, _your_ favorite workspaces should show up first (that is, favorite workspaces of other users should be sorted as they would if they were not favorited at all).

**Approach:**

The 'obvious' approach is to add a boolean column `favorite` to the workspaces table. This will work for one user, but then all users' favorites go to the top when filtering by multiple users. I decided to avoid this approach unless absolutely required.

Instead, I first tried adding a separate table `user_favorites` in which we store `(user_id, workspace_id)` tuples and adding some joins to return the favorite status by user. This proved unwieldy for `GetWorkspaceByID` as that changes the return signature to a `GetWorkspaceByIDRow` and breaks everything.

I then changed tactics and instead added a nullable UUID column `favorite_of`. This stores the UUID of the workspace _owner_ when marked as 'favorite', and is `NULL` otherwise. We can then pass in the UUID of the caller in order to control the ordering of 'favorites'.

We convert the `Favorite` status to a boolean value when converting between the respective `database` and `codersdk` structs.

I originally considered co-opting the `owner` parameter for this ordering but instead elected to add a separate parameter to avoid surprises (e.g. querying for `owner:larry` as `curly` would then show Larry's favorite workspaces first).

This means that:
- You must be able to _write_ a workspsace in order to mark it as favorite.
- Marking someone else's workspace as favorite will still make it show up as _their_ favorite, and not yours.

### CODE REVIEW

File: coderd/database/migrations/000186_user_favorite_workspaces.up.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE ONLY workspaces
+ADD COLUMN favorite_of uuid DEFAULT NULL;
+COMMENT ON COLUMN workspaces.favorite_of IS 'FavoriteOf contains the UUID of the workspace owner if the workspace has been favorited.';
```
Comment (johnstcn) [inline coderd/database/migrations/000186_user_favorite_workspaces.up.sql:3]: self-review: just storing a boolean value here doesn't provide enough context to the query to be able to know how to sort favorites.

**Review by mtojek — COMMENTED** [review]
> I then changed tactics and instead added a nullable UUID column favorite_of. This stores the UUID of the workspace owner when marked as 'favorite', and is NULL otherwise. We can then pass in the UUID of the caller in order to control the ordering of 'favorites'.

A comment on the design. Let's assume that Alice has workspace, and Bob is admin, but would like to mark Alice's workspace as favorite. Is it doable with this design?

**Comment by johnstcn** [comment]
> > I then changed tactics and instead added a nullable UUID column favorite_of. This stores the UUID of the workspace owner when marked as 'favorite', and is NULL otherwise. We can then pass in the UUID of the caller in order to control the ordering of 'favorites'.
> 
> A comment on the design. Let's assume that Alice has workspace, and Bob is admin, but would like to mark Alice's workspace as favorite. Is it doable with this design?

Bob can mark Alice's workspace as Alice's favorite on her behalf. Favoriting a workspace that you do not own isn't supported here, but is doable with some minor modifications:
- `favorite_of` becomes a `uuid[]`
- Add `user_id` to parameters of `FavoriteWorkspace` / `UnfavoriteWorkspace`

Is this something you think is worthwhile?

**Review by mtojek — COMMENTED** [review]
> Is this something you think is worthwhile?

I would say so, unless PMs have different preferences. Alice might be scared that somebody "hacked" her account and started interacting with "favorites" markers. 

> favorite_of becomes a uuid[]

I admit that I prefer to have a separate "favorites" table than expanding the workspaces table schema, but I'm not a database expert. I wouldn't like to mix concerns as there is always the risk of growing up the workspace table. You may want to consult our database guy @mafredri  :)

**Comment by johnstcn** [comment]
> > Is this something you think is worthwhile?
> 
> I would say so, unless PMs have different preferences. Alice might be scared that somebody "hacked" her account and started interacting with "favorites" markers.
> 

To be fair, the same can be said of any action an admin takes on a user's resources.

**Comment by johnstcn** [comment]
> I admit that I prefer to have a separate "favorites" table than expanding the workspaces table schema, but I'm not a database expert. I wouldn't like to mix concerns as there is always the risk of growing up the workspace table.

Yeah, this was my first attempt as well. Suffice it to say that adding any sort of join to the `GetWorkspaceByID` query will break many, many things and requires some annoying workarounds. Migrating this to a separate table down the line is definitely possible if required.

**Comment by mtojek** [comment]
> query will break many, many things and requires some annoying workarounds.

... unless you identify API calls that truly require "favorites" data. Maybe it is just the `GetWorkspaces` API call, so we can replace the database call with `GetWorkspacesWithShares`. Array field `favorites_of []uuid` could be an option too if we acknowledge risks above.

Anyway, I understand the complexity, and I don't intend to stop you. I'd like to save us from reimplementing this feature in the future if we make wrong product assumptions.

**Comment by Emyrk** [comment]
@mtojek Joining tables is more a limitation of our SQLc design. When you do any joins, the data structs become a mess.

So some of our database design, and likely this, is because of that. I think in a pure academic database setting, this could be normalized to another table. But the headache that gives up the stack makes a `[]uuid` column simpler imo. 

And table normalization would only really show value if we then use that table for some other queries somehow? Like "Show me the number of favorited workpaces for each template". Then the `[]uuid` column starts to fall short.

We can always normalize later though if we need, that migration doesn't feel too hard.

**Comment by mtojek** [comment]
> Like "Show me the number of favorited workpaces for each template". Then the []uuid column starts to fall short.

I'm afraid that devs could perceive it as an invitation to add more joined arrays, and build the single table model.

One more question, I might have missed it in the review, but will it be possible to expose the "favorite" information only to the requester? It would be useless to leak the information about who really "likes" the workspace.

File: coderd/database/queries/workspaces.sql [hunk 2]
```diff
@@ -262,6 +262,9 @@ WHERE
 	-- Authorize Filter clause will be injected below in GetAuthorizedWorkspaces
 	-- @authorize_filter
 ORDER BY
+	CASE WHEN workspaces.favorite_of = @order_by_favorite THEN
+		workspaces.favorite_of = @order_by_favorite
+	END ASC,
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/database/queries/workspaces.sql:267]: Interesting that you do not need an `ELSE` here. 

I think we can drop the body of the case, and just put `true`. Just to remove a bit of duplication.
```suggestion
	CASE WHEN workspaces.favorite_of = @order_by_favorite THEN
		true
	END ASC,
```

**Comment by johnstcn** [comment]
> > Like "Show me the number of favorited workpaces for each template". Then the []uuid column starts to fall short.
> 
> I'm afraid that devs could perceive it as an invitation to add more joined arrays, and build the single table model.
> 
> One more question, I might have missed it in the review, but will it be possible to expose the "favorite" information only to the requester? It would be useless to leak the information about who really "likes" the workspace.

I think this is covered by `convertWorkspace` taking the user ID of the requestor and only showing it as 'favorite' if the database struct `favorite_of` field matches the requestor's ID.

**Comment by mtojek** [comment]
Ok, thanks for clarifying my doubts. Feel free to pick the implementation you prefer.

File: coderd/workspaces.go [hunk 3]
```diff
@@ -1021,6 +1032,80 @@ func (api *API) putExtendWorkspace(rw http.ResponseWriter, r *http.Request) {
...
+	aReq.Old = workspace
+
+	err := api.Database.FavoriteWorkspace(ctx, workspace.ID)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error setting workspace as favorite",
+			Detail:  err.Error(),
+		})
+		return
+	}
```
Comment (Emyrk) [inline coderd/workspaces.go:1064]: It feels weird not to verify that the caller is the person. I understand this is so an admin can favorite a workspace for someone else, but still feels odd :thinking:.

Should we just not allow anyone to favorite someone else's workspace for now? If a feature request is made, we can add it back.

**Review by Emyrk — APPROVED** [review]
Left some nits, I think we should drop allowing admins to favorite other workspaces for now.

My nits are not blocking.

File: coderd/workspaces.go (hunk 3 above)
Comment (johnstcn) [reply coderd/workspaces.go:1064]: I'm happy to do that, it does smell like RBAC creep though.

**Comment by Emyrk** [comment]
Some notes for posterity:

Sorting by booleans is weird:

```
SELECT * FROM (VALUES(true), (false), (null)) s(a) ORDER BY a DESC;

| a     |
| ----- |
| null  |
| true  |
| false |
```

```
SELECT * FROM (VALUES(true), (false), (null)) s(a) ORDER BY a ASC;

| a     |
| ----- |
| false |
| true  |
| null  |
```

There is `ORDER BY x NULLS FIRST/LAST`, but it is better to just remove the possibility of nulls.

File: coderd/workspaces.go (hunk 3 above)
Comment (Emyrk) [reply coderd/workspaces.go:1064]: We do a lot of this outside rbac already unfortunately :cry:. This stuff doesn't fit in rbac atm.

File: coderd/database/queries/workspaces.sql [hunk 4]
```diff
@@ -262,9 +262,9 @@ WHERE
 	-- Authorize Filter clause will be injected below in GetAuthorizedWorkspaces
 	-- @authorize_filter
 ORDER BY
-	CASE WHEN workspaces.favorite_of = @order_by_favorite THEN
-		workspaces.favorite_of = @order_by_favorite
-	END ASC,
+	-- COALESCE because the result of the comparison is NULL if not true
+	-- and this messes up the ordering.
+	COALESCE(workspaces.favorite_of = @order_by_favorite, false) DESC,
```
Comment (Emyrk) [inline coderd/database/queries/workspaces.sql:267]: :+1:

**Comment by johnstcn** [comment]
This has gotten a bit big and unwieldy so I'm going to close this out and open a new PR with the feedback from this one.

**Pull request closed without merging**